### PR TITLE
Warn on missing parentheses in bitstring modifiers

### LINKED ISSUE #11811: Warn on variables becoming function calls in more occasions
There are some scenarios where a variable may become a function call:

  * [ ] `<<foo::binary()>>` should become `<<foo::binary>>`
  * [ ] `<<foo::custom_type>>` should become `<<foo::custom_type()>>`

Before we warn, we should first change the formatter to handle those cases. One complication is that the formatter does not change the AST, so we will have to say "in principle it does not, unless in very few cases, which can be disabled via an option". A new option must be added:

  * [x] `normalize_bitstring_modifiers: true` (#11840)

### PR DESCRIPTION
**Pull request by sabiwara**
Closes #11811

### CODE REVIEW

File: lib/elixir/test/elixir/kernel/warning_test.exs
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/warning_test.exs:1659]: Can you please add the tests to kernel/binary_test.exs?

**Review by josevalim — COMMENTED** [review]
Thank you! Can you also please warn if someone writes `<<foo::binary()>>` instead of `<<foo::binary>>`?

Note though that we can only merge this after v1.14 is out, because it is too early to introduce the warning. :)

**Comment by sabiwara** [comment]
> Thank you! Can you also please warn if someone writes `<<foo::binary()>>` instead of `<<foo::binary>>`?

Sure, will do!

> Note though that we can only merge this after v1.14 is out, because it is too early to introduce the warning. :)

Noted :)

File: lib/elixir/src/elixir_bitstring.erl
Comment (sabiwara) [inline lib/elixir/src/elixir_bitstring.erl:391]: I based this message on the following for consistency:
```
extra parentheses on a remote function capture &~ts.~ts()/0 have been deprecated. Please remove the parentheses: &~ts.~ts/0
 ```

File: lib/elixir/src/elixir_quote.erl
Comment (sabiwara) [inline lib/elixir/src/elixir_quote.erl:398]: Without this change, `@bitstring <<"foo", 16::4>>` was generating a warning because expanded to `binary()`.
Technically, this is a breaking change, but I think it is needed for consistency?

File: lib/elixir/test/elixir/kernel/binary_test.exs
Comment (sabiwara) [inline lib/elixir/test/elixir/kernel/binary_test.exs:267]: Note: With `defmacrop`, I wasn't able to access the custom modifiers in `Code.eval_string` even by passing `__ENV__`.
Please let me know if there is a better way.

**Review by josevalim — APPROVED** [review]
Looks great!

Waiting for the main branch to move to Elixir v1.15 before merging.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**
